igc.cc

```diff
@@ -400,12 +450,85 @@ void IgcFormat::read()
+       * We don't particularly care about that. After that, every group of seven
+       * bytes is 4 digits followed by three letters, specifying start end end
+       * bytes of each extension, and the kind of extension (always three chars)
+       *
+       * First, construct a hash of all available extensions.
+       * It may be better to construct the map and parse the record at the same time.
+       * However, for now we are contructing the hash map first, then iterating through it.
+       * The values in this hash contain all the necessary information on various extensions.
+      */
+      // TODO: Return this hash map for later use
+      QHash<QString, short>::const_iterator i;
+      if (global_opts.debug_level >= 1) {
+        printf(MYNAME ": I record: %s\n", ibuf_q.toUtf8().constData());
+      }
+      for (unsigned i=3; i < ibuf_q.length(); i+=7) {
+        QString ext_type = ibuf_q.mid(i+4, 3);
+        if (global_opts.debug_level >= 1) {
+          printf(" %s;",ext_type.toUtf8().constData());
+        }
+        if (ext_type == "ENL") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_enl);
+        } else if (ext_type == "TAS") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_tas);
+        } else if (ext_type == "VAT") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_vat);
+        } else if (ext_type == "OAT") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_oat);
+        } else if (ext_type == "TRT") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_trt);
+        } else if (ext_type == "GSP") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_gsp);
+        } else if (ext_type == "FXA") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_fxa);
+        } else if (ext_type == "GFO") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_gfo);
+        } else if (ext_type == "SIU") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_siu);
+        } else if (ext_type == "ACZ") {
+          ext_types_hash.insert(ibuf_q.mid(i, 7), ext_rec_acz);
+        }
+      }
+      if (global_opts.debug_level >= 1) {
+        printf("\n");
+      }
+      /*
+       * Now we have a hash list. The values of that list contain the extension
+       * type and begin and end bytes of each extension definition. Iterate through
+       * the hash and parse individual hash values.
+      */
+      if (global_opts.debug_level >= 2) {
+        printf(MYNAME ": Extnensions defined in I record:\n");
+        printf(MYNAME ": Extensions present:\n(Note: IGC I records "
+                      "are one-initialized. Strings are zero-initialized.)\n");
+      }
+      for (i = ext_types_hash.constBegin(); i != ext_types_hash.constEnd(); ++i) {
+        short begin = QStringView(i.key().mid(0,2)).toInt() - 1;
+        short end = QStringView(i.key().mid(2,2)).toInt() - 1;
+
+        QStringView ext_record_type = QStringView(i.key().mid(4,3)).toString();
```

Comment:
Ok, here we are using the keys.  But we still aren't using the hash to find values from keys => we don't need a hash.  We could use a QList<std::pair<QString, short>>.

`QStringView(i.key().mid(0,2)).toInt() - 1;`
unnecessarily constructs a QStringView.  just convert the QString that mid constructed.
short begin = i.key().mid(0,2).toInt() - 1;

`QStringView ext_record_type = QStringView(i.key().mid(4,3)).toString();`
i.key().mid(4,3) is a QString, which we convert to a QStringView, which we convert back to a QString, which we convert back to a QStringView!
instead:
QString ext_record_type = i.key().mid(4,3);

Reply:
@PacketFiend , please see the attached implementation of these ideas
[igc.patch](https://github.com/GPSBabel/gpsbabel/files/11128175/igc.patch)